System/Atom.cpp

```diff
@@ -0,0 +1,1060 @@
+
+					// Now calculate the total time left to travel, according to the progress made.
+					timeLeft -= timeLeft * segProgress;
+
+					// Move position forward to the hit position.
+					//position += segTraj * segProgress;
+					// Only move the dom forward by int domSteps, so we don't cross into a pixel too far
+					position[dom] += (domSteps + static_cast<int>(sinkHit)) * increment[dom];
+
+					// Move the submissive direction forward by as many int steps, or the full float segTraj if all sub-steps are clear
+					if ((subSteps + static_cast<int>(subStepped && sinkHit)) < delta[sub]) {
+						position[sub] += (subSteps + static_cast<int>(subStepped && sinkHit)) * increment[sub];
+					} else {
+						position[sub] += segTraj[sub];
+					}
+
+					Vector testPos = position - m_Offset;
+
+					didWrap = g_SceneMan.WrapPosition(testPos) || didWrap;
+
+					// Apply the collision response acceleration to the linear velocity of the owner MO.
+					velocity += hitAccel;
+				}
+			}
+		} while ((hit[X] || hit[Y]) && /* !segTraj.GetFloored().IsZero() && */ hitCount < 100 && !m_LastHit.Terminate[HITOR]);
+
+		//RTEAssert(hitCount < 100, "Atom travel resulted in more than 100 segments!!");
+
+		// Draw the trail
+		if (g_TimerMan.DrawnSimUpdate() && m_TrailLength) {
+			int length = m_TrailLength /* + 3 * PosRand()*/;
+			for (int i = trailPoints.size() - MIN(length, trailPoints.size()); i < trailPoints.size(); ++i) {
+				putpixel(trailBitmap, trailPoints[i].first, trailPoints[i].second, m_TrailColor.GetIndex());
+			}
+		}
+
+		// Unlock all bitmaps involved.
+		//if (m_TrailLength) { trailBitmap->UnLock(); }
+		if (!scenePreLocked) { g_SceneMan.UnlockScene(); }
+
+		// Extract Atom offset.
+		position -= m_Offset;
+
+		// Travel along the remaining segTraj.
+		if (!(hit[X] || hit[Y]) && autoTravel) { position += segTraj; }
+
+		didWrap = g_SceneMan.WrapPosition(position) || didWrap;
+
+		ClearMOIDIgnoreList();
+
+		return hitCount;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void HitData::Clear() {
+		HitPoint.Reset();
+		VelDiff.Reset();
+		BitmapNormal.Reset();
+		HitDenominator = 0;
+
+		for (unsigned short i = 0; i < 2; ++i) {
+			Body[i] = 0;
+			RootBody[i] = 0;
+			HitVel[i].Reset();
+			TotalMass[i] = 0;
+			MomInertia[i] = 0;
+			HitRadius[i].Reset();
+			HitMaterial[i] = 0;
+			PreImpulse[i].Reset();
+			ResImpulse[i].Reset();
+			ImpulseFactor[i] = 0;
+			SquaredMIHandle[i] = 0;
+			Terminate[i] = false;
+		}
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	HitData & HitData::operator=(const HitData &rhs) {
+		if (this == &rhs) {
+			return *this;
+		}
+		Clear();
+
+		HitPoint = rhs.HitPoint;
+		VelDiff = rhs.VelDiff;
+		BitmapNormal = rhs.BitmapNormal;
+		HitDenominator = rhs.HitDenominator;
+
+		for (unsigned short i = 0; i < 2; ++i) {
```

Comment:
I feel like there's some std thing you could do to avoid the loop. Though again, it doesn't really matter, this is fine.